### DISCUSSION TITLE
Virtual Tables

### DISCUSSION THREAD

**Discussion by danz68** [duckdb/duckdb#1674] [Ideas]
Duckdb is an embedded database engine. In order to use it, we need to populate somehow tables from reading external CSV's, Parquet or using Appender.  Using the Appender is a decent bridge between DuckDB engine and any sort of external managed data (btw, while seems to be thread safe, I see no sensible boost using 2, 4, 8 separate threads each with its own appender, but this is another story) . 

However, I believe that DuckDB engine can benefit from using a zero copy mechanism if a C program, for instance, would be ready to share some read only buffers as columns, avoiding duplication of data using the Appender mechanism. Data can be easily persisted on request using "insert from select", it can benefit from custom user compression mechanism, or custom connection drivers, without interfering with the DuckDB code. For this the engine needs to provide a mechanism for registration of the table virtual type. Ideally the projection pushdown and filters should be part of the fetching process.

Engine.RegisterVirtualTable("My Custom Type", &VirtualFetchFunction)

create table MyVirtualTable (A integer, B string, C string) virtual 'My Custom Type';

To simplify at maximum, the registred virtual table type should be able to provide a way of fetching data on column basis VirtualFetch("table name", "field name", optional_filter_as_bitset_or_record_ids_list)

A query like this: 
select A, B from MyVirtualTable where C='whatever';

Would invoke 3 times the fetching function:

VirtualFetch("table name", "field name", NULL);
... DuckDB does its magic computing the record filter, if any ...
VirtualFetch("table name", "A", optional_filter_as_bitset_or_record_ids_list);
VirtualFetch("table name", "B", optional_filter_as_bitset_or_record_ids_list);

Thoughts?

**Comment by Mytherin** [comment]
This suggestion is very similar to the already existing [Table Function interface](https://github.com/duckdb/duckdb/blob/6c7c9805fdf1604039ebed47d233ea55cabb4b2c/src/include/duckdb/function/table_function.hpp), which is what is used to enable scanning of e.g. Pandas DataFrames or Arrow tables in the Python client, and R DataFrames in the R client. In principle this is exposed to the user but we don't really have any user-facing documentation on that. We could remedy that and make usage of this API more clear to the user.

**Reply by danz68** [reply]
Even without documentation, once I've read the members description (as well as, briefly, parquet-extension as example) I can see already how exhaustive this solution can be. **Which is clearly good**. However, the way it is now, it may be used to extend DuckDB functionality using C++ only, and I think it has a lot of potential for the other API's as well. 

If possible (and with DuckDB it was) I prefer to not statically link the import library in an application, but to dynamically load every exported function of the library, using the C API header as a starting point. (BTW, an exported function returning the version of the engine would be helpful as well). I already miss in the C API the registration of scalar and vectorial functions, the TableFunction extension/development would be missed as well...

**Reply by Mytherin** [reply]
For now the C++ API is our main focus, as this is the language our extensions (and the library itself) are written in. We would  however more than welcome pull requests related to extending the C API with the functions you describe :)

**Reply by danz68** [reply]
I wish I was able to do so. Implementing duckdb in an application is one thing, but writing code for the library requires a different skill set and time availability. Unfortunately I can only make remarks about the flexibility of an embedded SQL engine, with UDF's and zero copy data manipulation features among them. 

(on a side note, "A in not distinct from B" not being implemented for merge joins is by design? 

select (select LKP.Field from LKP where LKP.JOIN_FIELD **is not distinct from** SELF.JOIN_FIELD) as LOOKUP_VALUE from SELF

seems to fail. In the meanwhile the old sql trick checking equality between coalesced terms still works, but the distinct approach would be the natural one. I know SQL purists would argue that we cant compare UNKNOWN values, but I would temper that saying that NULLs are used for EMPTY values as well, and those values can be compared, so distinct approach is very much a valid and welcome option. But, this is another discussion for another time... )
